add support for 1-D and 3-D images, Logger interface, TBImage Wrapper, Tests, Examples

shashikdm opened the pull request:
Added `L`, `CL`, `LC` formats. added respective tests also

Commit:
Merge pull request #1 from PhilipVinc/master

updating

Commit:
Merge pull request #2 from PhilipVinc/master

logger interface updates

Commit:
Merge pull request #5 from PhilipVinc/master

updating

Commit:
add matrix and list support to log_text

Commit:
reverting test_TBLogger.jl

Commit:
Merge pull request #7 from PhilipVinc/master

merge

Commit:
add support for 1-D images in `log_image`

oxinabox commented on src/Loggers/LogImage.jl:
```diff
@@ -1,11 +1,11 @@
 Log multiple images using `Array` of images and format
 - imgArrays: `Array` of images, e.g. Array{Array{Float64, 3}, 1}. `Array` of images can be multidimensional.
-- format: format which applies to each image in the `Array` of images. It can be one of {HW, WH, HWC, WHC, CHW, CWH, HWN, WHN, NHW, NWH, HWCN, WHCN, CHWN, CWHN, NHWC, NWHC, NCHW, NCWH}
+- format: format which applies to each image in the `Array` of images. It can be one of {L, CL, LC, HW, WH, HWC, WHC, CHW, CWH, HWN, WHN, NHW, NWH, HWCN, WHCN, CHWN, CWHN, NHWC, NWHC, NCHW, NCWH}
```
Can we have a link to some docs explaining what they mean?
Or a summary of what it means if it is simple?
I.eg I ma gessign maybe this is channels?
So C = Color? H= Height, W=Width, L=Length?, N=observation?

shashikdm replied:
okay i'll explain it in the docstring. that should do it? 
>C = Color? H= Height, W=Width, L=Length?, N=observation?  


this is correct

oxinabox replied:
Yep

Commit:
add ImageFormat explanation

Commit:
add seperate function for image objects

Commit:
add more formats
change flow of data
different `image_summary` for image object and AbstractArrays

Commit:
add preprocess function for image objects for automatic dispatch

Commit:
add TBImage, TBImages wrapper

Commit:
add tests. more required. more will come

Commit:
Merge pull request #8 from shashikdm/updateimage

Updateimage

shashikdm commented:
Highlights:
- seperate `image_summary` for {AbstractArray, format} and {AbstractArray{:<Colorant}}(image objects)
- autmatically detect format from image objects, must provide format for number arrays
- automatic dispatch of `image_summary` for image objects when using logger interface
- `TBImage` to be used only for numerical arrays and takes format as an argument
- `TBImages` also takes format as argument and creates TBImage objects for each

Commit:
minor bug fix

Commit:
Merge pull request #9 from shashikdm/updateimage

minor bug fix

shashikdm commented:
TODO:
- Refactoring
- more test coverage

Commit:
refactoring

Commit:
add support for 3-d images such as mri

shashikdm commented on src/Loggers/LogImage.jl:
```diff
@@ -18,21 +32,164 @@ end
+        #Grayscale/color matrix
+        height, width = size(img)
+    else
+        throw("Unknown Dimensions")
```
@oxinabox Please suggest what to throw here and at line 87. Perhaps throwing an appropriate exception instead of a string
at 87 I want to say I cannot handle/log this type of colorscheme.
at 97 I want to say too many dimensions for an Image (>3)

Commit:
bug fixes

Commit:
add more tests

Commit:
minor bug fix

Commit:
add examples folder

Commit:
Merge pull request #10 from shashikdm/examples

add examples folder

oxinabox commented on src/logger_dispatch_overrides.jl:
```diff
@@ -39,6 +39,7 @@ function preprocess(name, val::TBImages, data)
     n = 1
     for img in val.data
         preprocess(name*"/$n", TBImage(img, val.format), data)
+        n += 1
```
Just do `for (n, image) in enumerate(val.data)`

Commit:
syntax revision

Commit:
Merge branch 'master' of github.com:shashikdm/TensorBoardLogger.jl

oxinabox commented on src/Loggers/LogImage.jl:
```diff
@@ -1,38 +1,191 @@
+        colorspace = 6
+    else
+        throw("Unknown Colorspace")
+    end
```
This could be done by defining a method that dispatches:

like

```
colorspace_id(::Type{Gray}) = 1
colorspace_id(::Type{GrayA}) = 2
...
function image_summary(name::AbstractString, img::AbstractArray{<:Colorant})
     ...
    colorspace = colorspace_id(eltype(img))
    ...
end
```

oxinabox commented:
I was trying to workout how to cut down some of the common code  in `image_summary`
More iformation can be encoded into  the enum

``` 
# Note: we use the 100s place of the enum to represent the numer of dimensions
# And the 10s place to represent which dimension is the observation (N), 0 for no obseration dimension 
@enum ImageFormat begin
    L=100; 
    CL=200; LC; HW; WH; NL=210; LN=220;  
    HWC=300; WHC; CHW; CWH; NCL=310; NHW; NWH; NLC; CLN=330; LCN; HWN; WHN;  
    NHWC=410; NWHC; NCHW; NCWH; HWCN=440; WHCN; CHWN; CWHN; 
end

expected_ndims(x::ImageFormat) = Int(x) ÷ 100  # get the 100's digit
obs_dim(x::ImageFormat) = (Int(x) % 100) ÷ 10   # get the 10's digit
```

Commit:
change if else to function dispatch `image_summary`

shashikdm reviewed (commented):
> I was trying to workout how to cut down some of the common code in `image_summary`
> More iformation can be encoded into the enum
> 
> ```
> # Note: we use the 100s place of the enum to represent the numer of dimensions
> # And the 10s place to represent which dimension is the observation (N), 0 for no obseration dimension 
> @enum ImageFormat begin
>     L=100; 
>     CL=200; LC; HW; WH; NL=210; LN=220;  
>     HWC=300; WHC; CHW; CWH; NCL=310; NHW; NWH; NLC; CLN=330; LCN; HWN; WHN;  
>     NHWC=410; NWHC; NCHW; NCWH; HWCN=440; WHCN; CHWN; CWHN; 
> end
> 
> expected_ndims(x::ImageFormat) = Int(x) ÷ 100  # get the 100's digit
> obs_dim(x::ImageFormat) = (Int(x) % 100) ÷ 10   # get the 10's digit
> ```

this is awesome! thank you very much

Commit:
major revision `log_image` smart use of ImageFormats

shashikdm commented:
shorter code, only about 1/3rd of the original. Greate idea @oxinabox.

Commit:
change throw message

oxinabox commented on src/Loggers/LogImage.jl:
```diff
@@ -116,211 +139,15 @@ function image_summary(name::AbstractString, imgArray::AbstractArray, format::Im
+    HWC => img ->  permutedims(img, (3, 1, 2)),
+    WHC => img ->  permutedims(img, (3, 2, 1)),
+    CHW => img ->  img,
+    CWH => img ->  permutedims(img, (1, 3, 2))
```
I don't think the dict of functions gives much that a chain of terneries wouldn't
(and the ternaries would be faster)
```
img = 
	format == L   ? reshape(img, (1, 1, size(img, 1))) :
	format == CL  ? reshape(img, (size(img, 1), 1, size(img, 2))) :
	format == LC  ? reshape(transpose(img), (size(img, 2), 1, size(img, 1))) :
	format == HW  ? reshape(img, (1, size(img, 1), size(img, 2))) :
	format == WH  ? reshape(transpose(img), (1, size(img, 2), size(img, 1))), :
	format == HWC ? permutedims(img, (3, 1, 2)) : 
	format == WHC ? permutedims(img, (3, 2, 1)) :
	format == CHW ? img : 
	format == CWH ? permutedims(img, (1, 3, 2)) :
	#== else ==#    img
```

shashikdm replied:
I just tested with `@test`, ternaries are 8x faster

oxinabox commented on src/Loggers/LogImage.jl:
```diff
@@ -1,5 +1,20 @@
+strip_obs = Dict(
+NL=>L, LN=>L, NCL=>CL, CLN=>CL, NLC=>LC, LCN=>LC,
+NHW=>HW, HWN=>HW, NWH=>WH, WHN=>WH,
+NHWC=>HWC, NWHC=>WHC, NCHW=>CHW, NCWH=>CWH, HWCN=>HWC, WHCN=>WHC, CHWN=>CHW, CWHN=>CWH
```
Indentation.
lines 12, 14, and 15 should be indented.

oxinabox commented on src/Loggers/LogImage.jl:
```diff
@@ -1,5 +1,20 @@
+
+expected_ndims(x::ImageFormat) = Int(x) ÷ 100 # get the 100's digit
+obs_dim(x::ImageFormat) = (Int(x) % 100) ÷ 10 # get the 10's digit
+strip_obs = Dict(
```
this should be `const`.

Also idk if it is worth doing this, but this is a good use-case for a OrderedCollections.jl `LittleDict`
would would be `const strip_obs = freeze(LittleDict(`

It could also be done with a function of course
```
function strip_obs(format)
	if format in (NL, LN)
		L
	elseif format in (NCL, CLN, NLC, LCN)
		LC
	else if ...
	else 
		format
	end
end
```

Another nice option might be MLStyle.jl
https://thautwarm.github.io/MLStyle.jl/latest/syntax/pattern/#Or-Pattern-1

```
function strip_obs(format)
	@match format begin
	NL || LN => L
	NCL || CLN || NLC || LCN => LC
	...
	_ => format	
	end
end
```

shashikdm replied:
`const strip_obs` and `function strip_obs` both show very similar performance. using `@time`. Other options are appealing but then they'll add to dependencies. I'll stick with `const strip_obs` dictionary here.

Commit:
change dict to ternary. minor revision

zenna commented:
Just wanted to add that this is dope and highly useful for the upcoming NIPS deadline.

PhilipVinc committed:
Merge branch 'master' into master

PhilipVinc commented:
Very nice!
I can merge it anytime you want. 

I'll also tag another minor release with that.

shashikdm commented:
@PhilipVinc you can go ahead and merge this. I suggest waiting a few more days for `log_audio` and then tag a release. It's almost ready in my branch

PhilipVinc commented:
Thanks a lot for your efforts.

I'll tag a release anyway as we can just tag a new one after `log_audio` is implemented.
